# **Business Understanding**
About data:*
This is a transactional data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail

*Note:*

InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 
StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.	
InvoiceDate: Invoice Date and time. Numeric, the day and time when each transaction was generated.
UnitPrice: Unit price. Numeric, Product price per unit in sterling.
CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal, the name of the country where each customer resides

*Source:*
Chen, D. (2015). Online Retail [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5BW33.
[source](https://archive.ics.uci.edu/dataset/352/online+retail)

Loading libraries

In [16]:
import pandas as pd
import matplotlib.pyplot as plt

Loading data

In [40]:
df = pd.read_excel("data\online + retail\Online Retail.xlsx")

# **Data Understanding**
- can the user understand the data
- can the user see a summary of the data?  "storytell"
- shape
- datatypes
- number of misising values
- outliers
- descriptive stats(most important)


Overview of the data

In [27]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [21]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France


In [19]:
rows, columns =df.shape
output = f"{rows} records  and {columns} features"
print(output)

541909 records  and 8 features


In [30]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

In [ ]:
df.isnull().sum() / len(df) * 100 #returns the percentage of missing values in each column

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

- there exists invoices for ghost customers (null customers)

In [27]:
#calcculate the total purchase amouont per customer =>
df["totalAmountSpent"] = df["Quantity"] * df["UnitPrice"]

In [28]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID,totalAmountSpent
count,541909.000000,541909,541909.000000,406829.000000,541909.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570,17.987795
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000,-168469.600000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000,3.400000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000,9.750000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000,17.400000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000,168469.600000
std,218.081158,NaN,96.759853,1713.600303,378.810824


In [37]:
purchase_573585_df = df[df['InvoiceNo'] == "573585"]]

SyntaxError: unmatched ']' (4147649695.py, line 1)

In [39]:
clean_df["InvoiceNo"] = clean_df["InvoiceNo"].astype(str) 

In [38]:
purchase_573585_df

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: InvoiceNo, Length: 541909, dtype: bool

In [35]:
df.describe(include="object") #returns the summary statistics for categorical columns

C:\Users\USER\AppData\Local\Temp\ipykernel_12252\1375198864.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object") #returns the summary statistics for categorical columns


,InvoiceNo,StockCode,Description,Country
count,541909,541909,540455,541909
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


Insights:
- Average quantity of sales through out the period was at 10 units 
- 25% of the quantities sold were less than 1 
- 50% of the quanttities sold were lesss than 3 
- 75% of the quantities sold were less than 10
 
 *totalAmountSpent(sterlingpounds)*
- avg total income was 17.987795 
- 25% of the total client expence wre less than 3.4
- 50% of the totla client expence weer less than 9.75
- 75% of the total client expence were less than 17.4

In [31]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


Insights:
* 541909 records and 8 types
* the data is composed of both and numerical and categorical information together with date time
* the data contains missing values 

In [ ]:
clean_df = df.drop("CustomerID", axis=1)

In [6]:
clean_df = df.drop("CustomerID", axis=1)

In [7]:
clean_df.describe()#numerical data

,Quantity,InvoiceDate,UnitPrice
count,541909.000000,541909,541909.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114
min,-80995.000000,2010-12-01 08:26:00,-11062.060000
25%,1.000000,2011-03-28 11:34:00,1.250000
50%,3.000000,2011-07-19 17:17:00,2.080000
75%,10.000000,2011-10-19 11:27:00,4.130000
max,80995.000000,2011-12-09 12:50:00,38970.000000
std,218.081158,NaN,96.759853


In [40]:
difference = clean_df['InvoiceDate'].max() - clean_df['InvoiceDate'].min()
print(difference)

373 days 04:24:00


In [41]:
df[df["CustomerID"]  == "John Doe" ] #here we filter rows whose ustiomer ID is John Doe

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


In [45]:
df_r = clean_df[clean_df['Country'] == "Canada"]
df_r.head(2)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
119191,546533,20886,BOX OF 9 PEBBLE CANDLES,12,2011-03-14 13:53:00,1.95,Canada
119192,546533,79030D,"TUMBLER, BAROQUE",6,2011-03-14 13:53:00,1.65,Canada


In [ ]:
"""
- get the invoice number
- check if first letter is 'c' or it the invoce starts with c
- if it start we 'c' we can alert the team
"""

Simulating cancelled invoice since the data does not contain cancelled invoices 

In [13]:
clean_df.at[0, 'InvoiceNo'] = 'cc536365' #we have changed a cell value, whereby the cell is an intersection between a row and a column. The first parameter is the row index and the second parameter is the column name. 

In [14]:
clean_df.head(2)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,cc536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,United Kingdom


In [ ]:
clean_df["InvoiceNo"] = clean_df["InvoiceNo"].astype(str) #we changed the data type of the column InvoiceNo to string. This is because we want to check if the first letter of the invoice number is 'c' or not. If it is 'c' we can alert the team.

In [47]:
clean_df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
Country                   str
dtype: object

In [48]:
clean_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,United Kingdom


In [ ]:
"""
Simualting a scenario that implements an inbuild datatype class without using it
e.g. since we have the startswith() method in python strings what does it look like if we do not use it? 
simualtion done below: 

"""
# cancel_state = lambda x : x.startswith("c") # this is an anomyous function
# name = "Compton"
# name = name.lower()

# print(cancel_state(name))

'\nSimualting a scenario that implements an inbuild datatype class without using it\ne.g. since we have the startswith() method in python strings what does it look like if we do not use it? \nsimualtion done below: \n\n'

In [57]:
clean_df.describe()

,Quantity,InvoiceDate,UnitPrice
count,541909.000000,541909,541909.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114
min,-80995.000000,2010-12-01 08:26:00,-11062.060000
25%,1.000000,2011-03-28 11:34:00,1.250000
50%,3.000000,2011-07-19 17:17:00,2.080000
75%,10.000000,2011-10-19 11:27:00,4.130000
max,80995.000000,2011-12-09 12:50:00,38970.000000
std,218.081158,NaN,96.759853


Insights  
**Quantity:**
* Majority of quantities selected were on avrage 9 items
* the lowst quanityt of items picked were 1
* the highest number was 80995**
* 25% of the order quanitties had 1 order
* 50% of the order quanitties were 3 and below
* 75% of the order quantities were 10 and below
* there existed a spread/ deviation from the mean by 218 items

**InvoiceDate**
* 50% of the invoics were captured during 2011-07-19 17:17:00 this was on atuesday
* the first purchase/ invoice was done in 2010-12-01 08:26:00 this was on a Wednesday
* the last purchase/ invoice was done in 2011-12-09 12:50:00 this was on a Friday


In [58]:
fifty_percent_sales_df = clean_df[(clean_df["InvoiceDate"] == "2011-07-19 17:17:00") &  (clean_df["Quantity"] >= 3)]
fifty_percent_sales_df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
270892,560601,84596B,SMALL DOLLY MIX DESIGN ORANGE BOWL,3,2011-07-19 17:17:00,0.42,United Kingdom
270893,560601,84596F,SMALL MARSHMALLOWS PINK BOWL,3,2011-07-19 17:17:00,0.42,United Kingdom
270920,560601,84818,DANISH ROSE PHOTO FRAME,4,2011-07-19 17:17:00,0.79,United Kingdom


In [59]:
seventy_five_percent_sales_df = clean_df[(clean_df["InvoiceDate"] == "2011-10-19 11:27:00") &  (clean_df["Quantity"] >= 10)]
seventy_five_percent_sales_df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,Country
406428,571779,23319,BOX OF 6 MINI 50'S CRACKERS,24,2011-10-19 11:27:00,2.49,United Kingdom
406429,571779,22155,STAR DECORATION RUSTIC,12,2011-10-19 11:27:00,0.42,United Kingdom
406434,571779,22090,PAPER BUNTING RETROSPOT,10,2011-10-19 11:27:00,2.95,United Kingdom
406436,571779,35964,FOLKART CLIP ON STARS,12,2011-10-19 11:27:00,0.39,United Kingdom
406437,571779,22142,CHRISTMAS CRAFT WHITE FAIRY,10,2011-10-19 11:27:00,1.45,United Kingdom
406438,571779,84012,MAGIC SHEEP WOOL GROWING FROM PAPER,12,2011-10-19 11:27:00,0.85,United Kingdom
406439,571779,22586,FELTCRAFT HAIRBAND PINK AND BLUE,12,2011-10-19 11:27:00,0.85,United Kingdom
406440,571779,11001,ASSTD DESIGN RACING CAR PEN,16,2011-10-19 11:27:00,1.69,United Kingdom
406442,571779,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,48,2011-10-19 11:27:00,0.29,United Kingdom
406444,571779,21829,DINOSAUR KEYRINGS ASSORTED,36,2011-10-19 11:27:00,0.21,United Kingdom


In [60]:
clean_df.describe(include='object') #this will give us the description of the object data types in the dataframe

C:\Users\USER\AppData\Local\Temp\ipykernel_12428\2557817453.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  clean_df.describe(include='object') #this will give us the description of the object data types in the dataframe


,InvoiceNo,StockCode,Description,Country
count,541909,541909,540455,541909
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


# **Data Preparation**

# **Modelling**

# **Evaluation**

# **Deployment**